### PR TITLE
[export] Update dynamo_graph_capture_for_export to return GraphModule.

### PR DESCRIPTION
**Pull request by zhxchen17**
Make dynamo_graph_capture_for_export return a more compatible GraphModule object which is closer the the original behavior of dynamo

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

### CODE REVIEW

File: torch/fx/graph.py [hunk 1]
```diff
@@ -917,6 +917,42 @@ def _format_annotations(self, free_vars: list[str], expanded_def: bool) -> str:
         else:
             return "\n    " + "".join(x + "; " for x in has_annotation) + "\n"
 
+    def gen_var_bindings(self, fn_args, free_vars, expanded_def) -> str:
```
Comment (tugsbayasgalan) [inline torch/fx/graph.py:933]: Is it just codemod?

File: torch/export/_trace.py [hunk 2]
```diff
@@ -1486,12 +1486,10 @@ def _strict_export(
 
     orig_arg_names = gm_torch_level.graph._codegen.pytree_info.orig_args  # type: ignore[attr-defined]
 
-    gm_torch_level.graph._codegen = _PyTreeCodeGen(
-        _PyTreeInfo(
-            orig_arg_names,
-            gm_torch_level._in_spec,
-            out_spec,
-        )
+    gm_torch_level.graph._codegen.pytree_info = _PyTreeInfo(
```
Comment (tugsbayasgalan) [inline torch/export/_trace.py:1540]: Hmm why did it change?

File: test/export/test_experimental.py [hunk 3]
```diff
@@ -439,6 +438,22 @@ def make_inputs():
...
+
+        class MyModel(torch.nn.Module):
+            def forward(self, x):
+                y = x[:-1, :]  # [s0 - 1, 32]
+                stacked = torch.stack([y] * N, dim=0)  # [N * (s0 - 1), 32]
+                reshaped = stacked.reshape(-1, N, 32)  # [(s0 - 1), N, 32]
+                return reshaped
+
+        inps = (torch.randn(10, 32),)
+        ep = dynamo_graph_capture_for_export(MyModel())(*inps)
```
Comment (tugsbayasgalan) [inline test/export/test_experimental.py:501]: Can you also print out the graph here as well?

File: test/distributed/tensor/test_dtensor_export.py [hunk 4]
```diff
@@ -323,7 +323,7 @@ def test_annotate_aot_export_joint_with_descriptors_alone(self):
         [
             (
                 graph_capture_and_aot_export_joint_with_descriptors_v2,
-                "[[4, 10], [4], [10, 4], [10], [4, 10], [4], [10, 4], [10], [s64, 10], [s64, 10]]",
+                "[[4, 10], [4], [10, 4], [10], [s49, 10], [s49, 10]]",
```
Comment (tugsbayasgalan) [inline test/distributed/tensor/test_dtensor_export.py]: Why did this change?

File: torch/_dynamo/functional_export.py [hunk 5]
```diff
@@ -538,47 +551,73 @@ def backend_dummy(*example_inputs):
...
             get_metrics_context(),
             dynamo_timed("fullgraph_capture"),
         ):
-            out = fullgraph_capture(mod, args, kwargs)
+            out = fullgraph_capture(
+                mod,
+                args,
+                kwargs,
+                constraints=constraints,
+                _is_export_deprecated_do_not_use=True,
```
Comment (tugsbayasgalan) [inline torch/_dynamo/functional_export.py]: I thought we wouldn't need this?

Comment (zhxchen17) [reply torch/_dynamo/functional_export.py]: We don't need this for msl, but we need this for strict export.

File: test/distributed/tensor/test_dtensor_export.py (hunk 4 above)
Comment (zhxchen17) [reply test/distributed/tensor/test_dtensor_export.py]: To make sure the behavior stay close to your current impl as much as possible.

File: torch/export/_trace.py (hunk 2 above)
Comment (zhxchen17) [reply torch/export/_trace.py:1540]: We will have different subclass called _ExportCodeGen here, so it's wrong to always assign _PyTreeCodegen here.

File: torch/fx/graph.py (hunk 1 above)
Comment (zhxchen17) [reply torch/fx/graph.py:933]: I added a subclass called _ExportCodegen which adds shuffling.

File: torch/_dynamo/functional_export.py (hunk 5 above)
Comment (zhxchen17) [reply torch/_dynamo/functional_export.py]: okay I think I want to tackle msl's graph capture first, so I will remove this.

File: test/export/test_experimental.py [hunk 6]
```diff
@@ -439,6 +439,95 @@ def make_inputs():
...
+            model = Model()
+
+        import torch.fx.traceback as fx_traceback
+
+        with fx_traceback.preserve_node_meta():
+            gm = dynamo_graph_capture_for_export(model)(*input_fn())
+
+        """
+        def forward(self, args_0, args_1):
+            _tree_leaf_0, _tree_leaf_1, _tree_leaf_2, = pytree.tree_leaves((self, args_0, args_1,))
```
Comment (tugsbayasgalan) [inline test/export/test_experimental.py:574]: I guess we don't have to address in this PR, but in export, we preserve the input names.

Comment (tugsbayasgalan) [reply test/export/test_experimental.py:574]: Wait i think the codegen is supposed to handle the input names, seems like something went wrong here?

**Review by tugsbayasgalan — COMMENTED** [review]
Looks mostly good. Pls get review from @anijain2305 as well.

File: test/export/test_experimental.py [hunk 7]
```diff
@@ -439,6 +439,95 @@ def make_inputs():
...
+            L_fw_in_ , L_bw_in_ , = self._in_shuffle_graph(_tree_leaf_0, _tree_leaf_1, _tree_leaf_2)
+            l_fw_in_ = L_fw_in_
```
Comment (tugsbayasgalan) [inline test/export/test_experimental.py:576]: nit: this feels little redundant, any chance this can be simplified?

File: test/export/test_experimental.py (hunk 6 above)
Comment (zhxchen17) [reply test/export/test_experimental.py:574]: no the names are expected here. We should be able to fix them separately for export later.

File: test/export/test_experimental.py [hunk 8]
```diff
@@ -439,6 +439,95 @@ def make_inputs():
         test_inputs = make_inputs()
         self.assertEqual(gm(*test_inputs), foo(*test_inputs))
 
+    def test_dynamo_graph_capture_closure(self):
+        from torch.export import Dim
+
+        N = 3
+
+        class MyModel(torch.nn.Module):
```
Comment (anijain2305) [inline test/export/test_experimental.py:492]: Can you add a closed over tensor? It seems you only have an integer here.

Comment (zhxchen17) [reply test/export/test_experimental.py:492]: Yeah good point. Added

File: test/export/test_experimental.py [hunk 9]
```diff
@@ -439,6 +439,107 @@ def make_inputs():
...
+                z = x + outer
+                y = z[:-1, :]  # [s0 - 1, 32]
+                stacked = torch.stack([y] * N, dim=0)  # [N * (s0 - 1), 32]
+                reshaped = stacked.reshape(-1, N, 32)  # [(s0 - 1), N, 32]
+                return reshaped
+
+        inps = (torch.randn(10, 32),)
+        ep = dynamo_graph_capture_for_export(MyModel())(*inps)
+        """
+        def forward(self, *flat_proxy_args):
```
Comment (tugsbayasgalan) [inline test/export/test_experimental.py]: Can you do self.assertExpectedInline on these changes?

File: torch/_dynamo/functional_export.py [hunk 10]
```diff
@@ -538,47 +551,83 @@ def backend_dummy(*example_inputs):
...
-        if not hasattr(compiled_mod, "meta"):
-            compiled_mod.meta = {}  # type: ignore[attr-defined]
-        if isinstance(compiled_mod.meta, dict) and "fake_mode" not in compiled_mod.meta:
-            compiled_mod.meta["fake_mode"] = out.backend_input.fake_mode
-        return compiled_mod
-    elif inspect.ismethod(mod):
-        return types.MethodType(pytree_call, mod.__self__)
-    else:
-        return pytree_call
+def normalize_graph_module(gm):
```
Comment (tugsbayasgalan) [inline torch/_dynamo/functional_export.py:588]: I am pretty sure this is not enough for export, but probably fine for now.

File: torch/_dynamo/functional_export.py [hunk 11]
```diff
@@ -538,47 +551,83 @@ def backend_dummy(*example_inputs):
...
+        pyt = pytreeify(out, mod, args, kwargs)
+
+        graph_module = pyt.graph_module
+        tree_leaf_names = [
+            graph_module.graph._graph_namespace.create_name(f"_tree_leaf_{i}", None)
+            for i in range(pyt.num_flat_args)
+        ]
+        graph_module.graph._codegen = _ExportCodeGen(
+            _PyTreeInfo(
+                argument_names(inspect.signature(mod), args, kwargs),
```
Comment (tugsbayasgalan) [inline torch/_dynamo/functional_export.py:622]: Do we actually use argument names in torch IR produced graph? If not, can we add this logic in the follow up PR? if the previous export behaviour is that it doesn't produce torch IR with correct user argument names, i am ok btw.

File: torch/_dynamo/functional_export.py [hunk 12]
```diff
@@ -538,47 +551,83 @@ def backend_dummy(*example_inputs):
...
+            graph_module._non_persistent_buffers_set = (
+                pyt.root._non_persistent_buffers_set.copy()
+            )
+        graph_module._in_spec = pyt.in_spec
+        graph_module._out_spec = pyt.out_spec
+        assert not hasattr(graph_module, "_in_shuffle_graph")
+        assert not hasattr(graph_module, "_out_shuffle_graph")
+        graph_module._in_shuffle_graph = pyt.in_shuffle_graph
+        graph_module._out_shuffle_graph = pyt.out_shuffle_graph
+        delattr(graph_module, "_param_name_to_source")
```
Comment (tugsbayasgalan) [inline torch/_dynamo/functional_export.py:646]: Hmm is this the issue in previous torch IR graph capture as well?

File: torch/_dynamo/functional_export.py [hunk 13]
```diff
@@ -538,47 +551,83 @@ def backend_dummy(*example_inputs):
...
+        graph_module.recompile()
+        graph_module.meta["module_call_specs"] = (
```
Comment (tugsbayasgalan) [inline torch/_dynamo/functional_export.py:648]: Just for my understanding, technically you don't need this right because you are anyway running the bytecode?

File: test/export/test_experimental.py (hunk 7 above)
Comment (zhxchen17) [reply test/export/test_experimental.py:576]: what do you mean specifically by redundant?

File: test/export/test_experimental.py (hunk 9 above)
Comment (zhxchen17) [reply test/export/test_experimental.py]: oki

File: torch/_dynamo/functional_export.py (hunk 10 above)
Comment (zhxchen17) [reply torch/_dynamo/functional_export.py:588]: yeah, let's tackle torch.export separately.

File: torch/_dynamo/functional_export.py (hunk 11 above)
Comment (zhxchen17) [reply torch/_dynamo/functional_export.py:622]: ok I can add a TODO here if you're ok with that.

File: torch/_dynamo/functional_export.py (hunk 12 above)
Comment (zhxchen17) [reply torch/_dynamo/functional_export.py:646]: I think in previous impl, you just transform and return a new graph module which doesn't have all the attached attributes like this.

File: torch/_dynamo/functional_export.py (hunk 13 above)
Comment (zhxchen17) [reply torch/_dynamo/functional_export.py:648]: technically I'm not using this. This is just for passing some basic torch.export tests.

**Review by tugsbayasgalan — APPROVED** [review]
